File: despensa_angular/src/app/entities/ingredient.ts
import {Food} from "./food";

export class Ingredient {

  private aliment: Food;
  private quantity: number;
  private quantity_type: string;
  private optional: boolean;
  private db_id: number ;


  constructor(aliment: Food, quantity: number, quantity_type: string, optional: boolean, db_id: number) {
    this.aliment = aliment;
    this.quantity = quantity;
    this.quantity_type = quantity_type;
    this.optional = optional;
    this.db_id = db_id;
  }


  get _aliment(): Food {
    return this.aliment;
  }

  set _aliment(value: Food) {
    this.aliment = value;
  }

  get _quantity(): number {
    return this.quantity;
  }

  set _quantity(value: number) {
    this.quantity = value;
  }

  get _quantity_type(): string {
    return this.quantity_type;
  }

  set _quantity_type(value: string) {
    this.quantity_type = value;
  }

  get _optional(): boolean {
    return this.optional;
  }

  set _optional(value: boolean) {
    this.optional = value;
  }

  get _db_id(): number {
    return this.db_id;
  }

  set _db_id(value: number) {
    this.db_id = value;
  }

  equals(other: Ingredient) {
    return this._aliment.equals(other._aliment) && this._db_id === other._db_id
  }

  static cast(ingredient: any) {
    console.log('[Ingredient] [cast] -> ' + JSON.stringify(ingredient))

    let {aliment, quantity, quantity_type, optional, db_id} = ingredient;
    optional = optional === 1;
    db_id = db_id? db_id: 0;
    aliment = Food.cast(aliment);

    return new Ingredient(aliment, quantity, quantity_type, optional, db_id);
  }

  static castAsArrayIngredients(ingredients: any[]) {
    console.log('[Ingredient] [castAsArrayIngredients] -> ' + JSON.stringify(ingredients))
    return ingredients.map(ingredient => {
      return this.cast(ingredient);
    });
  }
}

File: libs/src/despensa/objects/ingredient.py
from dataclasses import dataclass
from typing import Any

from despensa.objects.aliment import Aliment
from despensa.objects.bd_instance import BDInstance


@dataclass
class Ingredient(BDInstance):
    """Class which represents an ingredient: an aliment for a recipe with a quantity

     Attributes
    ----------
    aliment : Aliment
        the concrete aliment
    quantity : int
        quantity of the aliment for a recipe
    quantity_type : str
        the type of the quantity. f.e. "gr", "kg", "spoons"...
    """

    aliment: Aliment
    quantity: float
    quantity_type: str = ""
    optional: bool = False
    db_id: int = 0

    def simple_str(self) -> str:
        """Returns a simplified str of the ingredient

        :return: Simple description of the ingredient
        """
        return f"{self.aliment.name.title()}: {self.quantity} {self.quantity_type}" + (
            " (Optional)" if self.optional else ""
        )

    @staticmethod
    def from_json(json: dict[str, Any]) -> "Ingredient":
        return Ingredient(
            Aliment.from_json(json["aliment"]),
            json["quantity"],
            json["quantity_type"],
            json["optional"],
            json["db_id"],
        )

    def __eq__(self, other):
        """Overrides the default implementation"""
        if isinstance(other, Ingredient):
            return (
                self.aliment == other.aliment
                and self.quantity == other.quantity
                and self.quantity_type == other.quantity_type
                and self.optional == other.optional
            )
        return False
